### PR TITLE
[BXMSPROD-1289] getLatestArtifactFromRepository

### PR DESCRIPTION
**Pull request by Ginxo**
https://issues.redhat.com/browse/BXMSPROD-1289

related PRs:

- https://gitlab.cee.redhat.com/middleware/build-configurations/-/merge_requests/369
- https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/150
- https://github.com/kiegroup/kogito-pipelines/pull/150

### CODE REVIEW

**Review by almope — APPROVED** [review]
LGTM

File: vars/maven.groovy [hunk 1]
```diff
@@ -132,9 +135,16 @@ def uploadLocalArtifacts(String mvnUploadCredsId, String artifactDir, String rep
     }
 }
 
+def getLatestArtifactFromRepository(String repositoryUrl, String groupId, String artifactId) {
+    def groupIdArtifactId = "${groupId.replaceAll("\\.", "/")}/${artifactId}"
+    def metadata = new XmlSlurper().parse("${repositoryUrl}/${groupIdArtifactId}/maven-metadata.xml")
+    return metadata.versioning && metadata.versioning.latest ? metadata.versioning.latest.text() : null
```
Comment (radtriste) [inline vars/maven.groovy:141]: Could something like that work ?

```suggestion
    return metadata.versioning?.latest?.text()
```

File: test/vars/MavenSpec.groovy [hunk 2]
```diff
@@ -148,105 +151,123 @@ class MavenSpec extends JenkinsPipelineSpecification {
...
-            1 * getPipelineMock('sh')('zip -r artifacts .')
-            1 * getPipelineMock('sh')("curl --silent --upload-file artifacts.zip -u ${mvnUploadCreds} -v ${repoUrl}")
+        1 * getPipelineMock('usernameColonPassword.call')([credentialsId: mvnUploadCredsId, variable: 'kieUnpack']) >> mvnUploadCreds
+        1 * getPipelineMock('withCredentials')([mvnUploadCreds], _ as Closure)
+        1 * getPipelineMock('dir')(artifactDir, _ as Closure)
+        1 * getPipelineMock('sh')('zip -r artifacts .')
+        1 * getPipelineMock('sh')("curl --silent --upload-file artifacts.zip -u ${mvnUploadCreds} -v ${repoUrl}")
+    }
+
+    def "[maven.groovy] getLatestArtifactFromRepository OK"() {
```
Comment (radtriste) [inline test/vars/MavenSpec.groovy:249]: what about a test with nothing returned ?

File: vars/maven.groovy [hunk 3]
```diff
@@ -132,9 +135,16 @@ def uploadLocalArtifacts(String mvnUploadCredsId, String artifactDir, String rep
     }
 }
 
+def getLatestArtifactFromRepository(String repositoryUrl, String groupId, String artifactId) {
```
Comment (radtriste) [inline vars/maven.groovy:138]: I would name it `getLatestArtifactVersionFromRepository`

File: vars/maven.groovy (hunk 1 above)
Comment (Ginxo) [reply vars/maven.groovy:141]: I've added bu adapting test method

**Review by almope — APPROVED** [review]

File: test/vars/MavenSpec.groovy (hunk 2 above)
Comment (Ginxo) [reply test/vars/MavenSpec.groovy]: null test for both methods implemented

File: vars/maven.groovy (hunk 3 above)
Comment (Ginxo) [reply vars/maven.groovy]: done

File: vars/maven.groovy [hunk 4]
```diff
@@ -132,9 +135,22 @@ def uploadLocalArtifacts(String mvnUploadCredsId, String artifactDir, String rep
...
+def getLatestArtifactVersionFromRepository(String repositoryUrl, String groupId, String artifactId) {
+    def groupIdArtifactId = "${groupId.replaceAll("\\.", "/")}/${artifactId}"
+    def metadata = new XmlSlurper().parse("${repositoryUrl}/${groupIdArtifactId}/maven-metadata.xml")
+    return metadata.versioning?.latest?.text()
+}
+
+def getLatestArtifactVersionPrefixFromRepository(String repositoryUrl, String groupId, String artifactId, String versionPrefix) {
+    def groupIdArtifactId = "${groupId.replaceAll("\\.", "/")}/${artifactId}"
+    def metadata = new XmlSlurper().parse("${repositoryUrl}/${groupIdArtifactId}/maven-metadata.xml")
+    return metadata.versioning?.versions?.childNodes().collect{ it.text()}.findAll{it.startsWith(versionPrefix)}.max()
```
Comment (radtriste) [inline vars/maven.groovy:147]:
```suggestion
    return metadata.versioning?.versions?.childNodes().collect{ it.text() }.findAll{ it.startsWith(versionPrefix) }.max()
```

File: vars/maven.groovy [hunk 5]
```diff
@@ -132,9 +135,22 @@ def uploadLocalArtifacts(String mvnUploadCredsId, String artifactDir, String rep
     }
 }
 
+def getLatestArtifactVersionFromRepository(String repositoryUrl, String groupId, String artifactId) {
+    def groupIdArtifactId = "${groupId.replaceAll("\\.", "/")}/${artifactId}"
+    def metadata = new XmlSlurper().parse("${repositoryUrl}/${groupIdArtifactId}/maven-metadata.xml")
```
Comment (radtriste) [inline vars/maven.groovy:140]: maybe could be extracted as it is repeated in both methods

Comment (Ginxo) [reply vars/maven.groovy]: done, thanks!

**Review by radtriste — APPROVED** [review]

**Comment by radtriste** [comment]
Looks like CI is failing with 
```
org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: Scripts not permitted to use staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods getAt java.lang.Object java.lang.String
```
@mbiarnes can you have a look ? seems like new Jenkins is not authorizing this (and this could be a huge problem for Kogito pipelines)

**Comment by mareknovotny** [comment]
https://stackoverflow.com/questions/38276341/jenkins-ci-pipeline-scripts-not-permitted-to-use-method-groovy-lang-groovyobject ? @radtriste

**Comment by radtriste** [comment]
@mareknovotny This was working like a charm on RHBA, that is why I am invoking Michael on this :)
I know about how to correct that on Jenkins but I have no rights to do it ...

**Comment by mareknovotny** [comment]
@radtriste the way is changed as JCasC needs to change this in git repo instead of on master instance in web UI. So we need to find a way there not on the Web UI
see e.g. https://github.com/jenkinsci/configuration-as-code-plugin/issues/415

**Comment by radtriste** [comment]
ok the error did not happen into the Kogito nightly tests so I guess it is a failure due to the Jenkinsfile code ...
anyway, it is not related to this PR, so I guess we can merge this one @mareknovotny

**Pull request merged**